src/lib.rs

```diff
@@ -0,0 +1,291 @@
+        write!(
+            f,
+            "container: {}, key: {}, reason: {}. ",
+            self.container, self.key, self.reason
+        )
+    }
+}
+
+fn validate(payload: &[u8]) -> CallResult {
+    let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
+
+    match validation_request.extract_pod_spec_from_object() {
+        Ok(pod_spec) => {
+            if let Some(pod_spec) = pod_spec {
+                return validate_pod_spec(pod_spec);
+            }
+            // If there is not pod spec, just accept it. There is no data to be validated.
+            kubewarden::accept_request()
+        }
+        Err(_) => {
+            // We were forwarded a request we cannot unmarshal or
+            // understand, just accept it
+            kubewarden::accept_request()
+        }
+    }
+}
+
+/// Scan all containers in containers, init_containers and ephemeral_containers.
+/// Returns reject_request if any secret var was found in an env_var in any container or
+/// accept_request otherwise.
+fn validate_pod_spec(pod_spec: PodSpec) -> CallResult {
+    let mut findings: HashSet<SecretRejected> = HashSet::new();
+    let secret_scanner = SecretScannerBuilder::new().build();
+
+    for container in pod_spec.containers {
+        if let Some(secrets_in_container) =
+            scan_env_vars(container.env, &secret_scanner, container.name)
+        {
+            findings.extend(secrets_in_container);
+        }
+    }
+
+    if let Some(init_containers) = pod_spec.init_containers {
+        for container in init_containers {
+            if let Some(secrets_in_container) =
+                scan_env_vars(container.env, &secret_scanner, container.name)
+            {
+                findings.extend(secrets_in_container);
+            }
+        }
+    }
+
+    if let Some(ephemeral_containers) = pod_spec.ephemeral_containers {
+        for container in ephemeral_containers {
+            if let Some(secrets_in_container) =
+                scan_env_vars(container.env, &secret_scanner, container.name)
+            {
+                findings.extend(secrets_in_container);
+            }
+        }
+    }
+
+    if !findings.is_empty() {
+        return kubewarden::reject_request(
+            Some(format!(
+                "The following secrets were found in environment variables -> {}",
+                create_error_message(findings)
+            )),
+            None,
+            None,
+            None,
+        );
+    }
+
+    kubewarden::accept_request()
+}
+
+fn scan_env_vars(
+    env_vars: Option<Vec<EnvVar>>,
+    secret_scanner: &SecretScanner,
+    container_name: String,
+) -> Option<HashSet<SecretRejected>> {
+    let mut findings: HashSet<SecretRejected> = HashSet::new();
+
+    if let Some(env_vars) = env_vars {
+        for env_var in env_vars {
+            if let Some(value) = env_var.value {
+                findings.extend(scan_env_var(
+                    value.as_bytes().to_vec(),
```

Comment:
I think we can change env_vars to be `Option<&[EnvVar]>` to avoid the conversion to `to_vec`

Reply:
I don't think that would help. We are converting to `to_vec` just the value, so we have a `Vec<u8>` for the [input](https://github.com/raulcabello/env-variable-secrets-scanner-policy/blob/main/src/lib.rs#L143)
Also `Option<Vec<EnvVar>>` is the [kubernetes type](https://docs.rs/k8s-openapi/0.7.1/k8s_openapi/api/core/v1/struct.Container.html#structfield.env), so we would need to convert it.